src/Common/tests/Collections/IListTest.cs

```diff
@@ -0,0 +1,1692 @@
+            {
+                var list = GetList(null);
+                object[] items;
+                if (ItemsMustBeNonNull)
+                {
+                    Assert.Throws<ArgumentNullException>(
+                        () => list.Add(null));
+                    items = GenerateItems(16);
+                    list.AddRange(items);
+                }
+                else
+                {
+                    items = new object[17];
+                    list.Add(null);
+                    items[0] = null;
+                    var tempItems = GenerateItems(16);
+                    list.AddRange(tempItems);
+                    Array.Copy(tempItems, 0, items, 1, 16);
+                }
+
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddWithNullValueAtEnd()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                var list = GetList(null);
+                object[] items;
+                var tempItems = GenerateItems(16);
+                list.AddRange(tempItems);
+
+                if (ItemsMustBeNonNull)
+                {
+                    Assert.Throws<ArgumentNullException>(
+                        () => list.Add(null));
+                    items = tempItems;
+                }
+                else
+                {
+                    items = new object[17];
+                    Array.Copy(tempItems, 0, items, 0, 16);
+                    list.Add(null);
+                    items[16] = null;
+                }
+
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddDuplicateValue()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                object[] items;
+                var list = GetList(null);
+                if (ItemsMustBeUnique)
+                {
+                    items = GenerateItems(16);
+                    list.AddRange(items);
+                    Assert.Throws<ArgumentException>(
+                        () => list.Add(items[0]));
+                }
+                else
+                {
+                    items = new object[32];
+                    var tempItems = GenerateItems(16);
+                    list.AddRange(tempItems);
+                    list.AddRange(tempItems);
+
+                    Array.Copy(tempItems, 0, items, 0, 16);
+                    Array.Copy(tempItems, 0, items, 16, 16);
+                }
+                CollectionAssert.Equal(items, list);
+            }
+        }
+
+        [Fact]
+        public void AddAndClear()
+        {
+            if (!ExpectedIsFixedSize && !ExpectedIsReadOnly)
+            {
+                var list = GetList(null);
+                list.AddRange(GenerateItems(16));
+                list.Clear();
+                var items = GenerateItems(16);
```

Comment:
Having it be the same length before/after the clear feels a little bit like cheating.

I'd probably go with smaller after.